ballerina/tests/from_json_test.bal

```diff
@@ -1264,3 +1264,346 @@ isolated function testRecordWithRestAsExpectedTypeForParseAsTypeNegative() {
+};
+
+type ReadonlyFieldsRec13 record {|
+    readonly int id;
+    readonly string name;
+    readonly string taxNo = "N/A";
+|};
+
+type ReadonlyFieldsRec14 record {
+    readonly int id;
+    readonly string name;
+    readonly string taxNo = "N/A";
+};
+
+type ReadonlyFieldsRec15 record {|
+    readonly int id;
+    readonly ReadonlyFieldsRec4 userDetails;
+    readonly string address = "N/A";
+|};
+
+type ReadonlyFieldsRec16 record {
+    readonly int id;
+    readonly ReadonlyFieldsRec4 userDetails;
+    readonly string address = "N/A";
+};
+
+type ReadonlyFieldsRec17 record {|
+    readonly int id;
+    readonly ReadonlyFieldsRec3 userDetails;
+    readonly string address = "N/A";
+|};
+
+type ReadonlyFieldsRec18 record {
+    readonly int id;
+    readonly ReadonlyFieldsRec3 userDetails;
+    readonly string address = "N/A";
+};
+
+type ReadonlyFieldsRec19 record {|
+    string name;
+    readonly int id;
+    readonly string taxNo = "N/A";
+|};
+
+type ReadonlyFieldsRec20 record {
+    readonly string taxNo = "N/A";
+    readonly int id;
+    string name;
+};
+
+
+@test:Config
+function testReadonlyFieldsWithDefaultValues() returns error? {
+    json user = {"id": 4012, "name": "John Doe"};
+    json user2 = {"id": 4012, "userDetails": user};
+    json user3 = {"id": 4012, "userDetails": user, taxNo: "1234", address: "Colombo", age: 19, name: "John Doe"};
+
+    ReadonlyFieldsRec11 r1 = check parseAsType(user);
+    test:assertEquals(r1, {id: 4012, taxNo: "N/A"});
+    test:assertEquals(r1.taxNo, "N/A");
+    
+    ReadonlyFieldsRec12 r2 = check parseAsType(user);
+    test:assertEquals(r2, {id: 4012, taxNo: "N/A", name: "John Doe"});
+    test:assertEquals(r2.taxNo, "N/A");
+    
+    ReadonlyFieldsRec13 r3 = check parseAsType(user);
+    test:assertEquals(r3, {id: 4012, taxNo: "N/A", name: "John Doe"});
+    test:assertEquals(r3.taxNo, "N/A");
+
+    ReadonlyFieldsRec14 r4 = check parseAsType(user);
+    test:assertEquals(r4, {id: 4012, taxNo: "N/A", name: "John Doe"});
+
+    ReadonlyFieldsRec15 r5 = check parseAsType(user2);
+    test:assertEquals(r5, {id: 4012, userDetails: {id: 4012, name: "John Doe"}, address: "N/A"});
+    
+    ReadonlyFieldsRec16 r6 = check parseAsType(user2);
+    test:assertEquals(r6, {id: 4012, userDetails: {id: 4012, name: "John Doe"}, address: "N/A"});
+    
+    ReadonlyFieldsRec17 r7 = check parseAsType(user2);
+    test:assertEquals(r7, {id: 4012, userDetails: {id: 4012, name: "John Doe"}, address: "N/A"});
+
+    ReadonlyFieldsRec18 r8 = check parseAsType(user2);
+    test:assertEquals(r8, {id: 4012, userDetails: {id: 4012, name: "John Doe"}, address: "N/A"});
+
+    ReadonlyFieldsRec19 r9 = check parseAsType(user);
+    test:assertEquals(r9, {id: 4012, taxNo: "N/A", name: "John Doe"});
+    
+    ReadonlyFieldsRec20 r10 = check parseAsType(user);
+    test:assertEquals(r10, {taxNo: "N/A", id: 4012, name: "John Doe"});
+
+    ReadonlyFieldsRec20 r11 = check parseAsType(user3);
```

Comment:
```suggestion
    ReadonlyFieldsRec20 r11 = check parseAsType(user3);
    r11.name = "New name";
```
Shall we update the noo-readonly attribute and assert it?

Reply:
Updated
https://github.com/ballerina-platform/module-ballerina-data.jsondata/pull/46/commits/754d86c4f043e042d320a1edcb08583c15d4c50c